resources/templates/md/posts/2016-04-04-state.md

```diff
@@ -0,0 +1,366 @@
+
+Greg has a series of different values (moves) over time (states 1-4).
+We need to realize, that we can transition the state but it is not possible to change or modify the previous state.
+The identity (Greg) can be modelled by using some reference type pointing to the value,
+and subsequently calling appropriate functions to create new states.
+
+NOTE:
+
+If you don't have any knowledge about threads in Clojure, you should read about [threads](http://blog.clojuregirl.com/clojure_threads.html) first.
+
+## How we can manage states (or how the data might change over time)
+
+There are four ways how to model an identity (manage states).
+Using reference types: `Vars` , `Atoms` , `Refs` , `Agents` .
+
+Greg has a girlfriend Gregorina and they go to the party. It's halloween party and
+there are a lot of sweets everywhere. Llamas love candies and as we know, 
+favourite candies for llamas are yummy marshmallows and gummy bears.
+
+![marshmallow](../img/marshmallow.png "Marshmallow")
+
+### Vars
+
+We can create a `Var` using `def`.
+In our example we create a `Var` named `marshmallows-in-the-bag` `(def marshmallows-in-the-bag)`
+and this `Var` will be reference for amount of the marshmallows in one bag.
+
+If we define `Var` as a `^:dynamic` (adding the `:^dynamic` metadata part).
+
+```clojure
+(def ^:dynamic marshmallows-in-the-bag 30)
+```
+
+`Var` can be rebound in different context and the rebinding is available only within it's dynamic scope.
+
+We will look what the dynamic scope is, and on the difference between Lexical and Dynamic scope.
+
+### Difference between the lexical and the dynamic scope
+
+What's the scope 
+
+Simply, the `scope` is a context created to summarize symbols. Without this scope all the symbols would be global.
+
+Let's start with the lexical scope
+
+### Lexical Scope
+
+Is the code block, where we define the symbol. This symbol is valid inside the code block where is defined (including any nested forms).
+The lexical scope can be defined for example by the `let` block.
+
+In the bag there are 30 marshmallows. So we will define `marshmallows-in-the-bag` with a value `30`.
+
+```clojure
+(def marshmallows-in-the-bag 30)
+```
+
+Now we will define a function `marshmallows-left`. This function takes one argument `marshmallows-eaten`. 
+In a let block we will define the `new-amount-of-marshmallows` and the value of this Var is the difference between `marshmallows-in-the-bag`
+and `marshmallows-eaten`. 
+As a last step in this function is the println `new-amount-of-marshmallows`.
+
+```clojure
+(defn marshmallows-left [marshmallows-eaten]
+  (let [new-amount-of-marshmallows (- marshmallows-in-the-bag marshmallows-eaten)]
+     (println (str "Marshmallows in the bag: " new-amount-of-marshmallows))))
+```
+
+Calling our function `marshmallows-left` with 10 (because Greg ate 10 pieces of marshmallow)
+
+```Clojure
+(marshmallows-left 10)
+
+;;=> Marshmallows in the bag: 20
+
```

Comment:
Overall layout: Check all of the examples for newlines like this. We can delete them:
<img width="524" alt="screen shot 2016-04-12 at 19 38 45" src="https://cloud.githubusercontent.com/assets/15321488/14469774/8a1f85b6-00e6-11e6-9bdd-f4670f4686dc.png">